src/DependencyGraph/ResolutionRequest.js

```diff
@@ -167,24 +171,21 @@ class ResolutionRequest {
 
           response.setResolvedDependencyPairs(mod, filteredPairs);
 
-          filteredPairs.forEach(([depName, modDep]) => {
-            p = p.then(() => {
-              if (!visited[modDep.hash()]) {
+          return Promise.all(
+            filteredPairs
+              .filter(([, modDep]) => !visited[modDep.hash()])
+              .map(([depName, modDep]) => {
                 visited[modDep.hash()] = true;
-                response.pushDependency(modDep);
-                if (recursive) {
-                  return collect(modDep);
-                }
-              }
-              return null;
-            });
-          });
-
-          return p;
+                return Promise.all([modDep, recursive ? collect(modDep) : []]);
```

Comment:
This is nice, I'm a bit concerned about I/O thrashing though because we'll fire off file-reads for every single JS file in the tree at the same time. In Jest on www, this completely stalled the process for about an hour (yes!). To mitigate this, I made a queue, like this: https://github.com/facebook/jest/blob/master/src/TestRunner.js#L203-L234

Another solution to limit concurrent jobs would be throat by our own @forbeslindesay: https://www.npmjs.com/package/throat

Do you think this will also be an issue here?